fish-rust/src/curses.rs

```diff
@@ -0,0 +1,300 @@
+//! A wrapper around the system's curses/ncurses library, exposing some lower-level functionality
+//! that's not directly exposed in any of the popular ncurses crates.
+//!
+//! In addition to exposing the C library ffi calls, we also shim around some functionality that's
+//! only made available via the the ncurses headers to C code via macro magic, such as polyfilling
+//! missing capability strings to shoe-in missing support for certain terminal sequences.
+//!
+//! This is intentionally very bare bones and only implements the subset of curses functionality
+//! used by fish
+
+use self::sys::*;
+use crate::common::ToCString;
+use std::cell::RefCell;
+use std::ffi::{CStr, CString};
+use std::sync::Arc;
+
+// This is `static mut` because a `Lazy` can only be init once but we may need to re-init the Term
+// value after a second+ call to curses::setup() in response to environment changes.
+pub static mut TERM: Option<Term> = None;
+
+/// Returns a reference to the global [`Term`] singleton or `None` if not preceded by a successful
+/// call to [`curses::setup()`].
+pub fn term() -> Option<&'static Term> {
+    unsafe { TERM.as_ref() }
+}
+
+/// Private module exposing system curses ffi.
+mod sys {
+    pub const OK: i32 = 0;
+    pub const ERR: i32 = -1;
+
+    extern "C" {
+        /// The ncurses `cur_term` TERMINAL pointer.
+        pub static mut cur_term: *const core::ffi::c_void;
+
+        /// setupterm(3) is a low-level call to begin doing any sort of `term.h`/`curses.h` work.
+        /// It's called internally by ncurses's `initscr()` and `newterm()`, but the C++ code called
+        /// it directly from [`initialize_curses_using_fallbacks()`].
+        pub fn setupterm(
+            term: *const libc::c_char,
+            filedes: libc::c_int,
+            errret: *mut libc::c_int,
+        ) -> libc::c_int;
+
+        /// Frees the `cur_term` TERMINAL  pointer.
+        pub fn del_curterm(term: *const core::ffi::c_void);
+
+        /// Checks for the presence of a termcap flag identified by the first two characters of
+        /// `id`.
+        pub fn tgetflag(id: *const libc::c_char) -> libc::c_int;
+
+        /// Checks for the presence and value of a number capability in the termcap/termconf
+        /// database. A return value of `-1` indicates not found.
+        pub fn tgetnum(id: *const libc::c_char) -> libc::c_int;
+
+        pub fn tgetstr(
+            id: *const libc::c_char,
+            area: *mut *mut libc::c_char,
+        ) -> *const libc::c_char;
+    }
+}
+
+pub struct Term {
+    /// A list of cached string values corresponding to known curses string capabilities. Individual
+    /// values may be overridden with [`Term::set()`].
+    ///
+    /// The list is prepopulated on initialization in [`Term::new()`], as such the `Vec` itself is
+    /// never manipulated - only potentially the values within its individual cells.
+    strings: Vec<RefCell<Option<Arc<CString>>>>,
```

Comment:
I think maybe where you and I are not on the same page is that *right now* the `Term` instance *is* prepopulated but **not immutably so**. As I mentioned previously, the way the C++ code sets up the terminal it is initialized in one place and at one point of time then the overrides are applied at another place at a later point in time. I *do* want to change that make it so that we have all our overrides available from the start to make it possible to just init an immutable `Term` (at least until the next one comes around), but that's not possible right now.

Your point about `RefCell` not being thread-safe is a good one but the `Arc` is there to allow the lifetime to extend past the `RefCell` borrow (in this case I believe the `RefCell` borrow is stripped out by the compiler that recognizes we only borrow to clone). I think the `RefCell` is safe to use because *in practice* we init and modify from the same thread at startup only, but that's not provable via the rust type system so I guess that means that we need to go back to a `Mutex` here after all.

The suggestion to use a boxed slice instead of a Vec is good; it makes it clearer that this structure is not intended to be manipulated and cannot be resized.